### PR TITLE
GH-33794: [Go] Add SetRecordReader to PreparedStatement 

### LINKED ISSUE #33794: [Go][FlightRPC] Add ability to bind a reader of parameters to Flight SQL prepared statement
### Describe the enhancement requested

This will let us bind a stream of parameters, not just a single batch.

This will be used to implement BindStream in the ADBC driver.

### Component(s)

Go

### PR DESCRIPTION
**Pull request by lidavidm**
### Are there any user-facing changes?

Adds a new method to PreparedStatement.
* Closes: #33794

### CODE REVIEW

File: go/arrow/flight/flightsql/client.go [hunk 1]
```diff
@@ -491,6 +490,33 @@ func (p *PreparedStatement) ExecuteUpdate(ctx context.Context) (nrecords int64,
...
+		}
+		return wr, nil
+	} else {
+		wr := flight.NewRecordWriter(pstream, ipc.WithSchema(p.streamBinding.Schema()))
+		wr.SetFlightDescriptor(desc)
+		for p.streamBinding.Next() {
+			if err := wr.Write(p.streamBinding.Record()); err != nil {
+				return nil, err
+			}
+		}
```
Comment (zeroshade) [inline go/arrow/flight/flightsql/client.go:512]: `arrio.Copy` ?

File: go/arrow/flight/flightsql/client.go [hunk 2]
```diff
@@ -518,22 +544,44 @@ func (p *PreparedStatement) GetSchema(ctx context.Context) (*flight.SchemaResult
...
+//
+// This will call Retain on the record to ensure it doesn't get released out
+// from under the statement. Release will be called on a previous binding
+// record or reader if it existed, and will be called upon calling Close on the
+// PreparedStatement.
+func (p *PreparedStatement) SetParameters(binding arrow.Record) {
+	p.clearParameters()
 	p.paramBinding = binding
 	p.paramBinding.Retain()
 }
```
Comment (zeroshade) [inline go/arrow/flight/flightsql/client.go:569]: should probably mention in the docstring that it will clear any existing bound params

Comment (zeroshade) [reply go/arrow/flight/flightsql/client.go:569]: nvm just saw you have `Release will be called on a previous binding record or reader`, ignore this :)

**Review by zeroshade — COMMENTED** [review]
LGTM just one idea to use `arrio.Copy` to simplify the reader/writer interaction.

File: go/arrow/flight/flightsql/client.go (hunk 1 above)
Comment (lidavidm) [reply go/arrow/flight/flightsql/client.go:512]: Unfortunately that takes a Reader, but streamBinding is a RecordReader

**Review by zeroshade — APPROVED** [review]
LGTM it's fine as is for now and we can look into arrio.Copy later. its not worth dealing with the interfaces to force it right now.

**Pull request merged**